python/ray/data/_internal/block_batching.py

```diff
@@ -119,74 +86,222 @@ def get_batches(block: Optional[ObjectRef[Block]] = None) -> Iterator[BatchType]
-            created.
-        n: The width of the sliding window.
-        clear_block_after_read: Whether to clear the leftmost block
-            from object store manually (i.e. without waiting for Python's
-            automatic GC) when it's out of the sliding window (i.e. been
-            consumed), so as to reclaim memory faster. The caller has to
-            ensure safety, i.e. the block will never be accessed again.
+        block_ref_iter: An iterator over block object references.
+        stats: Dataset stats object used to store block fetching time.
 
     Returns:
-        An iterator of n-width windows over iterable.
-        If n > len(iterable), then a single len(iterable) window is
-        returned.
+        An iterator over resolved blocks.
     """
-    it = iter(iterable)
-    window = collections.deque(itertools.islice(it, n), maxlen=n)
-    if len(window) > 0:
-        yield tuple(window)
-    for elem in it:
-        block_ref = window.popleft()
+
+    for block_ref in block_ref_iter:
+        if block_ref is not None:
+            stats_timer = stats.iter_get_s.timer() if stats else nullcontext()
+            with stats_timer:
+                block = ray.get(block_ref)
+            yield block
+
+
+def _prefetch_blocks(
+    block_ref_iter: Iterator[ObjectRef[Block]],
+    prefetcher: "BlockPrefetcher",
+    num_blocks_to_prefetch: int,
+    clear_block_after_read: bool = False,
+    stats: Optional[Union[DatasetStats, DatasetPipelineStats]] = None,
+) -> Iterator[ObjectRef[Block]]:
+    """Given an iterable of Block Object References, returns an iterator
+    over these object reference while prefetching `num_block_to_prefetch`
+    blocks in advance.
+
+    Args:
+        block_ref_iter: An iterator over block object references.
+        num_blocks_to_prefetch: The number of blocks to prefetch ahead of the
+            current block during the scan.
+        clear_block_after_read: Whether to clear the block from object store
+            manually (i.e. without waiting for Python's automatic GC) after it
+            is read. Doing so will reclaim memory faster and hence reduce the
+            memory footprint. However, the caller has to ensure the safety, i.e.
+            the block will never be accessed again.
+        stats: Dataset stats object used to store block wait time.
+    """
+
+    if num_blocks_to_prefetch == 0:
+        for block_ref in block_ref_iter:
+            yield block_ref
+            if clear_block_after_read:
+                ray._private.internal_api.free(block_ref, local_only=False)
+
+    window_size = num_blocks_to_prefetch
+    # Create the initial set of blocks to prefetch.
+    sliding_window = collections.deque(
+        itertools.islice(block_ref_iter, window_size), maxlen=window_size
+    )
+    with stats.iter_wait_s.timer() if stats else nullcontext():
+        prefetcher.prefetch_blocks(list(sliding_window.queue))
```

Comment:
Should be able to convert the top-level `deque` to a list like this; I actually don't think that `deque` has a `queue` attribute.
```suggestion
        prefetcher.prefetch_blocks(list(sliding_window))
```